proxy/src/proxy.rs

```diff
@@ -324,20 +324,42 @@ async fn connect_to_compute(
     extra: &console::ConsoleReqExtra<'_>,
     creds: &auth::BackendType<'_, auth::ClientCredentials<'_>>,
 ) -> Result<PostgresConnection, compute::ConnectionError> {
-    let mut num_retries: usize = NUM_RETRIES_WAKE_COMPUTE;
+    let mut num_retries: usize = 0;
     loop {
         // Apply startup params to the (possibly, cached) compute node info.
         node_info.config.set_startup_params(params);
         match connect_to_compute_once(node_info).await {
-            Err(e) if num_retries > 0 => {
-                info!("compute node's state has changed; requesting a wake-up");
-                match creds.wake_compute(extra).map_err(io_error).await? {
-                    // Update `node_info` and try one more time.
-                    Some(mut new) => {
-                        new.config.reuse_password(&node_info.config);
-                        *node_info = new;
+            Err(e) if num_retries < NUM_RETRIES_WAKE_COMPUTE => {
+                // not all errors are retryable
+                let wait_duration = match e {
+                    // retry all errors at least once
+                    _ if num_retries == 0 => Some(Duration::ZERO),
+                    // keep retrying connection errors every 100ms
+                    compute::ConnectionError::CouldNotConnect(_) => {
+                        Some(Duration::from_millis(100))
+                    }
+                    // otherwise, don't retry
+                    _ => None,
+                };
+
+                match wait_duration {
+                    Some(dur) => {
+                        info!("compute node's state has changed; requesting a wake-up");
+                        match creds.wake_compute(extra).map_err(io_error).await? {
```

Comment:
I think we only need to check with the console after the first failed connection attempt and no need to query it again each 100ms after that. So may be this code would be a bit simpler if the first attempt handled before the loop.

Reply:
Yeah that makes sense

Reply:
However, it is a bit more complicated. E.g. this trace https://neonprod.grafana.net/goto/wEHTMf94g?orgId=1. That `400 Bad Request: endpoint is in transition` from console is a retriable error, so if console request returned this error, then we should call wake_compute again in the retry code.

Reply:
Aha! I knew there would be a case where we would need to invalidate the cache again